modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/HashJoinNode.java

```diff
@@ -0,0 +1,623 @@
+                }
+            } else {
+                next = (Map<Object, Object>) next0;
+            }
+        }
+
+        return coll;
+    }
+
+    private static <RowT> List<RowT> getUntouched(Map<Object, Object> entries, @Nullable List<RowT> out) {
+        if (out == null) {
+            out = new ArrayList<>();
+        }
+
+        for (Map.Entry<Object, Object> ent : entries.entrySet()) {
+            if (ent.getValue() instanceof Collection) {
+                TouchedList<RowT> coll = (TouchedList<RowT>) ent.getValue();
+                if (!coll.touched) {
+                    out.addAll(coll);
+                }
+            } else {
+                getUntouched((Map<Object, Object>) ent.getValue(), out);
+            }
+        }
+        return out;
+    }
+
+    private static <RowT> void resetTouched(Map<Object, Object> entries) {
+        for (Map.Entry<Object, Object> ent : entries.entrySet()) {
+            if (ent.getValue() instanceof Collection) {
+                TouchedList<RowT> coll = (TouchedList<RowT>) ent.getValue();
+                if (coll.touched) {
+                    coll.touched = false;
+                }
+            } else {
+                resetTouched((Map<Object, Object>) ent.getValue());
+            }
+        }
+    }
+
+    @Override
+    protected void pushRight(RowT row) throws Exception {
+        assert downstream() != null;
+        assert waitingRight > 0;
+
+        checkState();
+
+        waitingRight--;
+
+        Map<Object, Object> next = hashStore;
+        int processed = 0;
+        for (Integer entry : rightJoinPositions) {
+            processed++;
+            Object ent = handler.get(entry, row);
+            if (processed == rightJoinPositions.size()) {
+                Collection<RowT> raw = touchResults
+                        ? (Collection<RowT>) next.computeIfAbsent(ent, k -> new TouchedList<>())
+                        : (Collection<RowT>) next.computeIfAbsent(ent, k -> new ArrayList<>());
+
+                raw.add(row);
+            } else {
+                next = (Object2ObjectOpenHashMap<Object, Object>) next.computeIfAbsent(ent, k -> new Object2ObjectOpenHashMap<>());
+            }
+        }
+
+        if (waitingRight == 0) {
+            rightSource().request(waitingRight = inBufSize);
+        }
+    }
+
+    void getMoreOrEnd() throws Exception {
+        if (waitingRight == 0) {
+            rightSource().request(waitingRight = inBufSize);
+        }
+
+        if (waitingLeft == 0 && leftInBuf.isEmpty()) {
+            leftSource().request(waitingLeft = inBufSize);
+        }
+
+        if (requested > 0 && waitingLeft == NOT_WAITING && waitingRight == NOT_WAITING && leftInBuf.isEmpty() && left == null
+                && !rightIt.hasNext()) {
+            requested = 0;
+            downstream().end();
+        }
+    }
+
+    private static class TouchedList<E> extends ArrayList<E> {
+        boolean touched = false;
+    }
```

Comment:
composition is better that inheritance in this case

Reply:
done